### ISSUE TITLE
Questions for the ODE function implementation 

### ISSUE THREAD

**Issue by zknus** [tk-rusch/GraphCON#6]
Hi team,

Thanks for your great work! I have one question regards your code https://github.com/tk-rusch/GraphCON/blob/3326c48bd3631b9f25417455ded9a1a83ccd3e93/src/homophilic_graphs/function_laplacian_diffusion.py#L55

when you implement the ODE functions for X' and Y', it seems different from the formula (8) in your paper.  The code of line 55 first half is Y', but what does the second half of $f$ meaning? Shouldn't X'=Y according to the formula (8)? 

Thanks!

**Comment by tk-rusch** [comment]
Hi,
This is essentially taken from the GRAND paper repository: https://github.com/twitter-research/graph-neural-pde/blob/main/src/function_laplacian_diffusion.py
It's simply a convex combination of alpha*x and the initial value of the ODE. This seems to work well in practice. 
Best,
Konstantin

**Comment by zknus** [comment]
> 



> Hi, This is essentially taken from the GRAND paper repository: https://github.com/twitter-research/graph-neural-pde/blob/main/src/function_laplacian_diffusion.py It's simply a convex combination of alpha*x and the initial value of the ODE. This seems to work well in practice. Best, Konstantin

Hi tk-rusch,

Thanks for your prompt response!  But I'm still confused, why do you use the concatenation here? Because in your ODE block final output, you only use the second half of this $f$. 
https://github.com/tk-rusch/GraphCON/blob/3326c48bd3631b9f25417455ded9a1a83ccd3e93/src/homophilic_graphs/GNN_early.py#L70-L75

And you already applied the  alpha*x and the initial value of the ODE to $f$  in 
https://github.com/tk-rusch/GraphCON/blob/3326c48bd3631b9f25417455ded9a1a83ccd3e93/src/homophilic_graphs/function_laplacian_diffusion.py#L53-L56

**Comment by tk-rusch** [comment]
Because it is a second-order ODE re-written as a system (or concatenation) of first-order ODEs. This can be seen in the example of a harmonic oscillator $y'' = -y$, which can be re-written with introducing $x=y'$ to $[x',y'] = [-y,x]$, where the brackets indicate concatenation. Does this help with your question?

Also note that we only use the position, i.e., X in our paper, for the final prediction. So we need the full concatenated system for solving the ODE, but for the final predictions we only need the positions X.

**Comment by zknus** [comment]
Hi tk-rusch,

Thanks for your explanation！
Yes, it's a second-order ODE and the concatenation means two first-order ODEs.  In the paper, the two first-order formulations is  $X'=Y$ and $Y'=\sigma(AX)-X-Y$ .  In the code, the first half of $f$ is $f_1=ay-y-x$ means the $Y'$ if we ignore the "add_source", while the second half is $(1-\beta) \alpha x + \beta x_0$ means the $X'$. However, why the implementation of $X'$ is  $(1-\beta) \alpha x + \beta x_0$ ? Shouldn't be   $(1-\beta) \alpha y + \beta y_0$ based on the formulation $X'=Y$ ?  Thanks！

**Comment by tk-rusch** [comment]
Okay now I understand the confusion: the roles of x and y are switched in this implementation compared to the notation in our paper. More concretely, in this implementation the $x$ is the $\bf Y$ in our paper and the $y$ in this implementation is the $\bf X$ in the paper. I implemented that before we wrote the paper and didn't bother to change it according to the notation in our paper.

**Comment by zknus** [comment]
Hi tk-rusch,

Now it makes sense! I still have a small question. Can you explain more about the "a convex combination of alpha*x and the initial value of the ODE" since you even implement this by $(1-\beta) \alpha x +  \beta x_0$ ?  Do you have any findings through the ablation study that if you don't add the initial value?  Thanks!

**Comment by tk-rusch** [comment]
No, we didn't ablate that (we only do it for Cora, Citeseer and pubmed, since we base these experiments on the GRAND paper implementation). Adding the initial value is quite standard -- we took that directly from the GRAND implementation. Adding it through a learnable convex combination was added by us, but that's also very standard (many gated RNNs work that way). 

Feel free to close this issue, if you don't have any more questions.
